### PR TITLE
Fix Issue 21050 - __traits(getOverloads) for templates returns incorr…

### PR DESCRIPTION
**Pull request by rmanthorpe**
…ect symbol for the first overload

### CODE REVIEW

**Comment by UplinkCoder** [comment]
There is no need for new AST node to support the desired effect as far as I can see.

**Comment by rmanthorpe** [comment]
Yeah I didn't really like this solution. What's the alternative approach? I could try to clone the `TemplateDeclaration` and null out `overnext`.

**Comment by UplinkCoder** [comment]
> What's the alternative approach? I could try to clone the TemplateDeclaration and null out overnext.

Sounds much better.

**Comment by BorisCarvajal** [comment]
> > What's the alternative approach? I could try to clone the TemplateDeclaration and null out overnext.
> 
> Sounds much better.

But that will create two different instances for the same template arguments.

For example this code should print "instantiated" only one time.
```
struct S {
    static void foo(T)(int i) { pragma(msg, "instantiated"); }
    static void foo(T)(string s) { }
}

alias foo0 = __traits(getOverloads, S, "foo", true)[0];
alias bar0 = foo0!long;  // instantiated here
void main()
{
    S.foo!long(0);  // already instantiated
}
```

**Comment by UplinkCoder** [comment]
@BorisCarvajal not necessarily.

But the concerns are valid, we'll have to see how it turns out, eventually a new flag like `cloned` has to be introduced for TemplateDeclaration.

**Comment by rmanthorpe** [comment]
This may sound naive but... What if I `syntaxCopy`, run semantic, then copy the `instances` field so they share the hash table of instantiations.

**Comment by UplinkCoder** [comment]
@rmanthorpe I have no idea if that would work, my assumption is that it violated some invariant, but try it.

**Comment by rmanthorpe** [comment]
I'll give it a go. How problematic is it if templates get instantiated more than once?

**Comment by UplinkCoder** [comment]
@rmanthorpe I can't say.
Maybe it'll do nothing bad, in the worst case it's silently disastrous.

**Comment by rmanthorpe** [comment]
This is at least a problem that a new AST node doesn't have, though it has others.

**Comment by rmanthorpe** [comment]
As far as I can work out any form of copying, even doing a `memcpy` and then nulling out `overnext`, results in duplicate template instances. The AST node approach does not. I've added a check for that to the test.

I think the only way to do this is going to involve having _something_ which has a pointer to the original `TemplateDeclaration`.  Either a new AST node, or adding a `cloned` field which contains the original.

**Comment by NilsLankila** [comment]
The overhaed when building dparse is ~3MB. This goes in the wrong direction as currently we try to pay attention to this.
Alas there might be no other solution. I have #11461, that would compensate this change.

**Comment by rmanthorpe** [comment]
@NilsLankila I'm not at all sure if this is the right way to go. I need to dig a bit deeper into where the multiple instantiations come from when I have time. A copy-without-overnext solution does seem like the best option if I can make it work.

**Comment by NilsLankila** [comment]
okay, maybe turn it as a draft to make clear that it's not readey even if it passes the tests.
Also you have the PR for the "child" traits that's almost good, it just needs a changelog entry.

**Comment by BorisCarvajal** [comment]
Creating an empty `TemplateDeclaration.instances` fixes the multiple instantiation problem.
Like this:
```
diff --git a/src/dmd/traits.d b/src/dmd/traits.d
index f6a7873a4..da6df2b6e 100644
--- a/src/dmd/traits.d
+++ b/src/dmd/traits.d
@@ -1081,7 +1081,28 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
                 if (!fd)
                 {
                     if(includeTemplates)
-                        exps.push(new DsymbolExp(Loc.initial, s, false));
+                    {
+                        if (auto td = s.isTemplateDeclaration())
+                        {
+                            if (td.instances is null)
+                            {
+                                // create an empty AA just to copy it
+                                scope ti = new TemplateInstance(Loc.initial, Id.empty, null);
+                                auto tib = TemplateInstanceBox(ti);
+                                td.instances[tib] = null;
+                                td.instances.clear();
+                            }
+                            auto oinstances = td.instances;
+                            td = cast(TemplateDeclaration)td.syntaxCopy(null);
+                            import core.stdc.string : memcpy;
+                            memcpy(cast(void*)td, cast(void*)s, __traits(classInstanceSize, TemplateDeclaration));
+                            // use the original instances table
+                            td.instances = oinstances;
+                            // make sure overnext is null to avoid overload set logic in the future
+                            td.overnext = null;
+                            exps.push(new DsymbolExp(Loc.initial, td, false));
+                        }
+                    }
                     return 0;
                 }
                 if (e.ident == Id.getVirtualFunctions && !fd.isVirtual())
```

**Comment by rmanthorpe** [comment]
> Creating an empty TemplateDeclaration.instances fixes the multiple instantiation problem.

@BorisCarvajal Thank you! You're totally right I feel like an idiot! I tried this but it didn't occur to me that `instances` might be null :facepalm:

File: src/dmd/traits.d [hunk 1]
```diff
@@ -1080,8 +1080,36 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
+                            {
+                                if (td.instances is null)
+                                {
+                                    // create an empty AA just to copy it
+                                    scope ti = new TemplateInstance(Loc.initial, Id.empty, null);
+                                    auto tib = TemplateInstanceBox(ti);
+                                    td.instances[tib] = null;
+                                    td.instances.clear();
+                                }
+                                auto oinstances = td.instances;
```
Comment (BorisCarvajal) [inline src/dmd/traits.d:1101]:
```suggestion
```
This line is not needed with `memcpy`. (I was originally using `dsymbolSemantic` instead and forgot to delete it)

File: src/dmd/traits.d [hunk 2]
```diff
@@ -1080,8 +1080,36 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
+                                td = cast(TemplateDeclaration) td.syntaxCopy(null);
+                                import core.stdc.string : memcpy;
+                                memcpy(cast(void*) td, cast(void*) s,
+                                        __traits(classInstanceSize, TemplateDeclaration));
+                                td.instances = oinstances;
```
Comment (BorisCarvajal) [inline src/dmd/traits.d:1106]:
```suggestion
```
ditto.

**Review by BorisCarvajal — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Review by BorisCarvajal — APPROVED** [review]

File: src/dmd/traits.d [hunk 3]
```diff
@@ -1080,8 +1080,34 @@ Expression semanticTraits(TraitsExp e, Scope* sc)
...
-                        exps.push(new DsymbolExp(Loc.initial, s, false));
+                    if (includeTemplates)
+                    {
+                        if (auto td = s.isTemplateDeclaration())
+                        {
+                            // if td is part of an overload set we must take a copy
+                            // which shares the same `instances` cache but without
+                            // `overroot` and `overnext` set to avoid overload
+                            // behaviour in the result.
+                            if(td.overnext !is null)
```
Comment (NilsLankila) [inline src/dmd/traits.d:1091]: there's a space missing between the keyword and the left parens.

Comment (rmanthorpe) [reply src/dmd/traits.d]: I'm not doing well at this. Fixed now thanks.

**Pull request merged**